detection_rules/rule_loader.py

```diff
@@ -4,45 +4,209 @@
+
+    def load_dict(self, obj: dict, path: Optional[Path] = None):
+        contents = TOMLRuleContents.from_dict(obj)
+        rule = TOMLRule(path=path, contents=contents)
+        self.add_rule(rule)
+
+        return rule
+
+    def load_files(self, paths: Iterable[Path]):
+        """Load multiple files into the collection."""
+        for path in paths:
+            self.load_file(path)
+
+    def _deserialize_toml(self, path: Path) -> dict:
+        # use pytoml instead of toml because of annoying bugs
+        # https://github.com/uiri/toml/issues/152
+        # might also be worth looking at https://github.com/sdispater/tomlkit
+        with io.open(str(path.resolve()), "r", encoding="utf-8") as f:
+            toml_dict = pytoml.load(f)
+            self._toml_cache[path] = toml_dict
+            return toml_dict
+
+    def _get_paths(self, directory: Path, recursive=True) -> List[Path]:
+        if recursive:
+            paths = glob.glob(str(directory / "**" / "*.toml"), recursive=True)
+        else:
+            paths = glob.glob(str(directory / "*.toml"))
+
+        return [Path(path) for path in paths]
+
+    def load_file(self, path: Path) -> TOMLRule:
         try:
-            return f(*args, **kwargs)
-        finally:
-            load_rules.clear()
+            path = path.resolve()
+
+            # use the default rule loader as a cache.
+            # if it already loaded the rule, then we can just use it from that
+            if self.__default is not None and self is not self.__default and path in self.__default.file_map:
+                rule = self.__default.file_map[path]
+                self.add_rule(rule)
+                return rule
+
+            obj = self._deserialize_toml(path)
+            return self.load_dict(obj, path=path)
+        except Exception:
+            print(f"Error loading rule in {path}")
+            raise
+
+    # noinspection PyShadowingBuiltins
+    def load_directory(self, directory: Path, recursive=True, toml_filter: Optional[Callable[[dict], bool]] = None):
+        paths = self._get_paths(directory, recursive=recursive)
+        if toml_filter is not None:
+            paths = [path for path in paths if toml_filter(self._deserialize_toml(path))]
 
-    return wrapped
+        self.load_files(paths)
 
+    def load_by_id(self, rule_id: definitions.UUIDString,
+                   directory: Path = Path(RULES_DIR), recursive=True) -> Optional[TOMLRule]:
+        """Search for a rule by ID and load it."""
+        if rule_id in self.id_map:
+            return self.id_map[rule_id]
+        elif self.__default is not None and rule_id in self.id_map:
+            return self.__default.id_map[rule_id]
 
-def reset():
-    """Clear all rule caches."""
-    load_rule_files.clear()
-    load_rules.clear()
-    get_rule.clear()
-    filter_rules.clear()
+        # otherwise, go looking
+        self.load_directory(directory, toml_filter=dict_filter(rule__rule_id=rule_id), recursive=recursive)
```

Comment:
so this method is called `load_by_id` but it really loads the full directory and then returns the rule by ID. It sort of feels like it could cause unintentional loading by misunderstanding the method name. I know it is for performance reasons since it makes sense to store them since they are loading anyway.

Should the onus be on the caller to ensure they run `load_directory` first so it is more explicit and then just do a simple lookup in the id_map?.

Also, to retrieve a rule directly from the collection, is the intent to use the filter method (even for a single rule) or to directly access `RuleCollection.id_map(id)`

Reply:
yeah loading by ID is the tricky part for sure.
i wonder if we can reduce the number of times we need to do this.

maybe a single CLI function that takes an ID and prints out a path could work.
then all CLI functions would take a path

Reply:
Looks like this is the only thing keeping me from approving.

An optimization we could make is to have a `rules-reference.json` which maps `rule.id` to `path` and then hook `TomlRule.save_toml` to check if the relative path is the default rules dir, and if so, always update it. Not a big fan of this because of manual moves

Reply:
yeah that seems overall worse.
maybe we should sync directly, so i can better understand the workflows where all you have is an ID

Reply:
I am fine with just doing
```python
rules = RuleCollection.default()
rule = rules.id_map.get(...)
```

So if anything just rename this `get_by_id` and expose the `id_map` or else just access `RuleCollection().id_map` directly

Reply:
so the downside with that is that every rule has to pass validation completely before you can look up by ID.

before we would load unvalidated toml and search.  so I was trying to do something similar: search through a bunch of TOML searching for the rule and only then perform the full marshmallow validation.

otherwise, there's a good chance it would blow up by trying to validate an unrelated rule and there'd be more overhead

Reply:
I would be good with the method as is, so long as it wasn't saving to the cache. After playing with it for a while, I think `RuleCollection` could benefit from having some of the functionality broken out a bit more.

<details>

```diff
--- a/detection_rules/rule_loader.py
+++ b/detection_rules/rule_loader.py
@@ -148,15 +148,32 @@ class RuleCollection:
     def _get_paths(self, directory: Path, recursive=True) -> List[Path]:
         return sorted(directory.rglob('*.toml') if recursive else directory.glob('*.toml'))
 
+    def _get_rule_from_default(self, path: Path):
+        # use the default rule loader as a cache.
+        # if it already loaded the rule, then we can just use it from that
+        if self.__default is not None and self is not self.__default and path in self.__default.file_map:
+            rule = self.__default.file_map[path]
+            self.add_rule(rule)
+            return rule
+
+    def _load_from_file_contents_multi(self, file_contents: Dict[Path, dict]):
+        for path, toml_dict in file_contents.items():
+            try:
+                path = path.resolve()
+
+                if rule := self._get_rule_from_default(path):
+                    return rule
+
+                return self.load_dict(toml_dict, path=path)
+            except Exception:
+                print(f"Error loading rule in {path}")
+                raise
+
     def load_file(self, path: Path) -> TOMLRule:
         try:
             path = path.resolve()
 
-            # use the default rule loader as a cache.
-            # if it already loaded the rule, then we can just use it from that
-            if self.__default is not None and self is not self.__default and path in self.__default.file_map:
-                rule = self.__default.file_map[path]
-                self.add_rule(rule)
+            if rule := self._get_rule_from_default(path):
                 return rule
 
             obj = self._deserialize_toml(path)
@@ -166,13 +183,20 @@ class RuleCollection:
             raise
 
     # noinspection PyShadowingBuiltins
-    def load_directory(self, directory: Path, recursive=True, toml_filter: Optional[Callable[[dict], bool]] = None):
+    def load_directory(self, directory: Path, recursive=True):
         paths = self._get_paths(directory, recursive=recursive)
-        if toml_filter is not None:
-            paths = [path for path in paths if toml_filter(self._deserialize_toml(path))]
-
         self.load_files(paths)
 
+    def _get_filtered_file_contents(self, directory: Path, toml_filter: Callable[[dict], bool], recursive=True):
+        paths = self._get_paths(directory, recursive=recursive)
+        file_contents = {path: toml_dict for path in paths if toml_filter(toml_dict := self._deserialize_toml(path))}
+        return file_contents
+
+    def load_filtered_directory(self, directory: Path, toml_filter: Callable[[dict], bool], recursive=True):
+        file_contents = self._get_filtered_file_contents(
+            directory=directory, toml_filter=toml_filter, recursive=recursive)
+        self._load_from_file_contents_multi(file_contents)
+
     def load_by_id(self, rule_id: definitions.UUIDString,
                    directory: Path = DEFAULT_RULES_DIR, recursive=True) -> Optional[TOMLRule]:
         """Search for a rule by ID and load it."""
@@ -181,11 +205,10 @@ class RuleCollection:
         elif self.__default is not None and rule_id in self.id_map:
             return self.__default.id_map[rule_id]
 
-        # otherwise, go looking
-        self.load_directory(directory, toml_filter=dict_filter(rule__rule_id=rule_id), recursive=recursive)
-
-        # now check if it was found
-        return self.id_map.get(rule_id)
+        # otherwise, go looking, without loading
+        rules = self._get_filtered_file_contents(directory, toml_filter=dict_filter(rule__rule_id=rule_id),
+                                                 recursive=recursive)
+        return rules.get(rule_id)
 
     def freeze(self):
         """Freeze the rule collection and make it immutable going forward."""
```

</details>

This breaks out a method to do just the load with no cache `self._get_filtered_file_contents` (which is shared with other methods - see below)

#### Double deserialization on filtered loads

While playing with this, I noticed that rules which are loaded with a filter in `load_directory` will be double loaded: first in `load_directory` and then again in `load_file` which calls `_deserialize_dict`

That is why you see some other extra functions within the diff above, namely `load_filtered_directory` being broken out from `load_directory`

Reply:
so load directory already has an optional filter, so you don't need two methods.
looks like i forgot to check `_toml_cache` in `load_file`

Reply:
still in general i don't really care for the approach where you load by ID, which requires a lot of this. it's worth syncing up to understand those use cases more, and why it's pervasive in so many CLI commands

Reply:
If there are concerns, get rid of it. There are other ways to achieve it without the method https://github.com/elastic/detection-rules/pull/1063/files#r601878748